lib/srv/db/common/auth.go

```diff
@@ -822,6 +833,47 @@ func (a *dbAuth) getCurrentAzureVM(ctx context.Context) (*libazure.VirtualMachin
 	return vm, nil
 }
 
+// GetAtlasIAMToken returns the AWS IAM token used to connect to MongoDB Atlas
+// instance.
+func (a *dbAuth) GetAtlasIAMToken(ctx context.Context, sessionCtx *Session) (string, string, string, error) {
+	dbAWS := sessionCtx.Database.GetAWS()
+	awsAccountID := dbAWS.AccountID
+
+	// If AWS account ID is empty, we need to fetch it from the agent identity.
+	// It has to be filled in case the database username is a partial ARN.
+	if awsutils.IsPartialRoleARN(sessionCtx.DatabaseUser) && awsAccountID == "" {
+		a.cfg.Log.Debugf("Fetching AWS Account ID to build role ARN")
+		stsClient, err := a.cfg.Clients.GetAWSSTSClient(ctx, dbAWS.Region)
+		if err != nil {
+			return "", "", "", trace.Wrap(err)
+		}
+
+		identity, err := awslib.GetIdentityWithClient(ctx, stsClient)
+		if err != nil {
+			return "", "", "", trace.Wrap(err)
+		}
+
+		awsAccountID = identity.GetAccountID()
+	}
+
+	arn, err := awsutils.BuildRoleARN(sessionCtx.DatabaseUser, dbAWS.Region, awsAccountID)
+	if err != nil {
+		return "", "", "", trace.Wrap(err)
+	}
+
+	sess, err := a.cfg.Clients.GetAWSSession(ctx, dbAWS.Region, cloud.WithAssumeRole(arn, sessionCtx.Database.GetAWS().ExternalID))
```

Comment:
What if the dbAWS.Region is empty ? 
I think that we support the case where a full ARN is provided and region is extracted from ARN.

Reply:
I've tested with an empty region, and it works fine. AFAIK the region is only used to determine the AWS partition (gov, china, or default); in that case, users can leave it empty if they are not on a different partition.

> I think that we support the case where a full ARN is provided and region is extracted from ARN.

The IAM role ARN doesn't contain the region, although it does have the partition. So I don't know if we should assume it from the agent ARN.

Reply:
Some UX questions:
1. should a mongo atlas database use fields from `aws` metadata block? I think this is fine. though not sure if UX is better if the mongo atlas section has its own `region`, `external_id` etc.
1. will mongo atlas support the aws cross-account feature?

Reply:
> I'm unsure if duplicating the configuration might be better since this AWS-specific integration could require something like aws_account_id and so on. So I'm not sure. However, we can improve the error messages and the documentation to mention which config users must fill.

👍 

> Yes, it does support (I've tested using a similar setup, except that the IAM role placed on the Atlas user was from an external account). However, I'm unsure how practical that would be since this integration requires no additional setup on the AWS account.

I don't see `dbAWS.AssumeRoleARN` being used here. Maybe I am missing something?

Reply:
> I don't see `dbAWS.AssumeRoleARN` being used here. Maybe I am missing something?

Is it required tho? In this case, we always assume the users' roles (from `tsh` command) as database usernames. However, it might require some change because it validates whether both `external_id` and `assume_role_arn` are provided.

I've tested with the following configuration, and it worked as expected. The teleport agent was using a role from account `111111111111`.
```yaml
- name: atlas
   protocol: "mongodb"
   uri: "mongodb+srv://tp.xxxxxxx.mongodb.net"
   tls:
     ca_cert_file: "isrgrootx1.pem"
   aws:
     account_id: "2222222222220"
     external_id: "example-external-id"
     assume_role_arn: "arn:aws:iam::222222222222:role/placeholder-role"
```

```code
$ tsh db connect --db-user=arn:aws:iam::222222222222:role/atlas-alice-role atlas
... success ...
```

Reply:
The agent assumes the `atlas-alice-role` directly. I've had to place the `assume_role_arn` in the configuration to fulfill the validation.

Reply:
When  `assume_role_arn` is to `placeholder-role`, it's expected that `placeholder-role` will assume `atlas-alice-role`. I would say it's more about role organization. Database agent assumes only the `placeholder-role` (most likely from a different AWS account) and `placeholder-role` has IAM permissions to other database access permissions like assuming `atlas-alice-role` in this case. This currently applies to all AWS databases.

I think we should either honour this logic for mongo atlas, or do not support `assume_role_arn` for atlas and return an error.

@gabrielcorado @smallinsky what do you think?

Reply:
@greedy52 
>  it's expected that placeholder-role will assume atlas-alice-role. 
Agree. we should support to role assumption chain. 

@gabrielcorado Could you provide a alignment for this ?

Reply:
I've updated it to do the role chaining.